client/service/tofu.go

```diff
@@ -0,0 +1,184 @@
+			}
+			if now.After(cert.NotAfter) {
+				_ = conn.Close()
+				return nil, ErrCertExpired
+			}
+		}
+		if !matched {
+			_ = conn.Close()
+			return nil, ErrNoMatchingFingerprint
+		}
+		// If we've gotten this far, then we can trust the server
+		slog.Debug("Server cert(s) passed TOFU tests")
+		return conn, nil
+	}
+	return &http.Client{
+		Transport: &http.Transport{
+			DialTLSContext: dial,
+		},
+	}, nil
+}
+
+// getServerFingerprints connects to an untrusted server, and get cert information about it
+// the addr must be a valid address, including port
+func getServerFingerprints(addr string) ([]CertDetail, error) {
+	var ret []CertDetail
+
+	// If the port isn't specified, assume https standard port
+	if !strings.Contains(addr, ":") {
+		addr = addr + ":443"
+	}
+	slog.Debug("Performing TOFU check", "addr", addr)
+
+	config := &tls.Config{
+		InsecureSkipVerify: true, // By definition, we're trying to build the trust...
+	}
+	conn, err := tls.Dial("tcp", addr, config)
+	if err != nil {
+		slog.Error("Unable to connect to", "addr", addr, "err", err)
+		return nil, err
+	} else {
+		state := conn.ConnectionState()
+
+		for _, cert := range state.PeerCertificates {
+			ret = append(ret, CertDetail{
+				Subject:     cert.Subject.CommonName,
+				Issuer:      cert.Issuer.CommonName,
+				Fingerprint: getFingerprint(cert.Raw),
+			})
+		}
+		if len(ret) == 0 {
+			return nil, ErrNoCertsDetected
+		}
+		return ret, nil
+	}
+}
+
+// TrustFingerprintCallback is a callback function that is called when a server's certificate
+// fingerprint is detected. It receives the server's IP address and the certificate details.
+// The responder can return a CertDetail to be used as the trusted fingerprint.
+// If the returned CertDetail is nil, the connection will be rejected.
+type TrustFingerprintCallback func(ip string, details []CertDetail) *CertDetail
+
+func readTrustedServerFingerprints(fingerprintsJSON string) (map[string]string, error) {
+	fingerprints := map[string]string{}
+	if _, err := os.Stat(fingerprintsJSON); err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return fingerprints, nil
+		}
+		return nil, fmt.Errorf("failed to check trusted server fingerprints: %w", err)
+	}
+
+	data, err := os.ReadFile(fingerprintsJSON)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read trusted server fingerprints: %w", err)
+	}
+
+	if err := json.Unmarshal(data, &fingerprints); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal trusted server fingerprints: %w", err)
+	}
+
+	return fingerprints, nil
+}
+
+func writeTrustedServerFingerprints(fingerprintsJSON string, fingerprints map[string]string) error {
+	data, err := json.Marshal(fingerprints)
+	if err != nil {
+		return fmt.Errorf("failed to marshal trusted server fingerprints: %w", err)
+	}
+	if err := os.MkdirAll(filepath.Dir(fingerprintsJSON), 0755); err != nil {
+		return err
+	}
+
+	if err := os.WriteFile(fingerprintsJSON, data, 0644); err != nil {
```

Comment:
SSH uses `0600` for permissions on `~/.ssh/authorized_keys`, which may be a good idea to use here as well.